include/simdjson/generic/ondemand/logger-inl.h

```diff
@@ -121,17 +155,21 @@ inline void log_headers() noexcept {
   }
 }
 
-inline void log_line(const json_iterator &iter, const char *title_prefix, const char *title, std::string_view detail, int delta, int depth_delta) noexcept {
-  log_line(iter, iter.position()+delta, depth_t(iter.depth()+depth_delta), title_prefix, title, detail);
+template <typename... Args>
+inline void log_line(const json_iterator &iter, const char *title_prefix, const char *title, std::string_view detail, int delta, int depth_delta, log_level level, Args&&... args) noexcept {
+  log_line(iter, iter.position()+delta, depth_t(iter.depth()+depth_delta), title_prefix, title, detail, level, std::forward<Args>(args)...);
 }
-inline void log_line(const json_iterator &iter, token_position index, depth_t depth, const char *title_prefix, const char *title, std::string_view detail) noexcept {
-  if (LOG_ENABLED) {
+
+template <typename... Args>
+inline void log_line(const json_iterator &iter, token_position index, depth_t depth, const char *title_prefix, const char *title, std::string_view detail, log_level level, Args&&... args) noexcept {
+  if (LOG_ENABLED && simdjson_unlikely(should_log(level))) {
```

Comment:
I am not sure you need the `simdjson_unlikely` there. Is there are reason to put it in?

In the scenario I am concerned about `LOG_ENABLED` is off, so this second part should never get evaluated in production.

Reply:
I think it should short circuit, but I could move it down as well just to be safe. 
```c++
if(LOG_ENABLED)
{
    if (simdjson_unlikely(should_log(level)))
    {
        // log...
    }
}
```
The reason for simdjson unlikely is to remove logging instructions from the hot path, and it may be helpful when logging is used rarely, in exceptional cases.

Should I remove the simdjson unlikely? I'll move the should log into another if clause just to be safe.

Reply:
I am not exactly sure I understand what you are saying..

> The reason for simdjson unlikely is to remove logging instructions from the hot path, and it may be helpful when logging is used rarely, in exceptional cases.

If logging is disabled, then LOG_ENABLED should be false.

We can leave the unlikely there, it is probably not harming us, but if you don't log, then you should not compile with LOG_ENABLED set to true. That's definitively not how we want you to use the library.

Reply:
I agree. `SIMDJSON_VERBOSE_LOGGING=0` would set `LOG_ENABLED` to false and should have the best performance.

There are 3 possible cases.

1. `SIMDJSON_VERBOSE_LOGGING=0`. Logging is completely disabled. This is the most performant, user pay no penalty.
2. `SIMDJSON_VERBOSE_LOGGING=1`, env `SIMDJSON_LOG_LEVEL=INFO` or unset. This is the least performant. This is likely used for debugging and performance isn't important.
3. `SIMDJSON_VERBOSE_LOGGING=1`, env `SIMDJSON_LOG_LEVEL=ERROR`. This performance is between that of 1st and 2nd case. In happy cases (no errors), no logs should appear. There should be no calls to `log_error` and the calls to `log_info` will do a load + compare instruction corresponding to `simdjson_unlikely(should_log(level))`, and that will evaluate to false (`log_threshold` is a singleton, so `getenv` only happens once).

I was referring to case 3. In happy cases, the user would only incur the penalty of some load and compare instructions, without incurring the full logging cost. By using `unlikely` it may help with instruction cache locality, the only instructions inserted into the hot paths are the compare instructions. But in exceptional cases when error happens, we have to pay the full cost of logging.